src/history/hs_rec.c

```diff
@@ -559,20 +523,13 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
             /* We can only delete history store entries that have timestamps. */
             WT_ERR(__wt_hs_delete_key_from_ts(session, btree->id, key, 1));
             WT_STAT_CONN_INCR(session, cache_hs_key_truncate_mix_ts);
-            clear_hs = false;
             F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
-        } else
-            /*
-             * Clear the content with timestamps in the history store if we see updates without
-             * timestamps on the update chain.
-             *
-             * We don't need to clear the history store records if everything is still on the insert
-             * list and there are no updates moved to the history store by checkpoint or a failed
-             * eviction.
-             */
-            clear_hs = first_non_ts_upd != NULL &&
-              !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS) &&
-              (list->ins == NULL || ts_updates_in_hs);
+        } else if (first_non_ts_upd != NULL && !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS) &&
+          (list->ins == NULL || ts_updates_in_hs)) {
+            WT_ERR(__wt_hs_delete_key_from_ts(session, btree->id, key, 1));
+            WT_STAT_CONN_INCR(session, cache_hs_key_truncate_mix_ts);
+            F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
```

Comment:
All callers to `__wt_hs_delete_key_from_ts` MUST call it before inserting updates with non-timestamped stat_ts's otherwise we can't guarantee ordering after insertion and will result in a visibility issue with respect to txnids.

Reply:
If I understand it correctly, we now always clear the history store before we do any insertion?